### ISSUE TITLE
Improve default configuration for `FunctionNaming`

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#5903]
Right now `FunctionNaming` doesn't check any test folder. But that's a bit too much. Not all the functions inside the test source set are test functions. You can have a lot of helper functions and those should follow the normal convention. More about that here: https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods

For that reason I think that we should do this change:

```diff
   FunctionNaming:
     active: true
-    excludes: ['**/test/**', '**/androidTest/**', '**/commonTest/**', '**/jvmTest/**', '**/androidUnitTest/**', '**/androidInstrumentedTest/**', '**/jsTest/**', '**/iosTest/**']
+    ignoreAnnotated: ['Test']
     functionPattern: '[a-z][a-zA-Z0-9]*'
     excludeClassPattern: '$^'
```

I know `@Test` sounds to junit and we don't want frameworks inside our default configuration. But let's be clear, junit is by far the most used framework to run tests. And if that doesn't sound like a good argument: The annotation `kotlin.test.Test` exists. And it is not even `kotlinx`. And it has exactly the same semantics as the junit `@Test` annotations. So, for that reason, I think that it would be sane for our documentation to do that change.

Also it would help with thinks like #5896 (but to make it clear, this issue is not because of that other one)

**Comment by cortinico** [comment]
> I know `@Test` sounds to junit and we don't want frameworks inside our default configuration

I'm a bit against this change for this specific reason you already mentioned, but I understand the rationale behind your proposal.

> But let's be clear, junit is by far the most used framework to run tests

We could argue that Dagger is by far the most used DI framework, and so on.

**Comment by BraisGabin** [comment]
@detekt/maintainers what do you think about this? I would like to hear more opinions.

The reason to close this is clear: framework related. And I would not argue against that. But I also would like you to consider about the annotation `kotlin.test.Test` (that I think is created to mimic the junit one) and that even in the [Kotlin Code Convention they talk about the `@Test` annotation](https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods). Could this be an execption? Or does this just make the things more complex?

**Comment by TWiStErRob** [comment]
TL;DR: I agree, with additional arguments **for** the change, and I would expand the scope.

I like the idea, in each of my projects I have to configure this, but I'm on the JVM & Android, what about KMP? What would you use in common tests, or iOS tests or web tests? I guess it's the `kotlin.test.Test` annotation, and therefore it's Kotlin specific, not framework specific.

I would argue that "we don't want frameworks inside our default configuration" fails when we add **any** hardcoded string, even just `**/test/**` couples the config with Gradle/Maven contentions. Also there's some coupling to Kotlin stdlib, coroutines, and Java (IgnoredReturnValue, ForbiddenAnnotation, ForbiddenAnnotation). The majority of users will use these by default, but some doesn't apply to KMP, and also there are alternative libs like Arrow, or even Guava, Apache Commons.

My worry is that this might be the same for other rules using all these test `excludes`, should we only change this one? e.g. just a very quick glance: UndocumentedPublicClass, MagicNumber, LateinitUsage, <I stopped looking at this point, because I think I proved the point>.

Also, should we first see if that default makes sense for us internally? e.g. this is overriding the defaults, but not in a way you describe in OP:
https://github.com/detekt/detekt/blob/main/detekt-gradle-plugin/config/gradle-plugin-detekt.yml

**Comment by BraisGabin** [comment]
I just tried and with this configuration our project works as a charm:

```yaml
  FunctionNaming:
    active: true
    excludes: []
    ignoreAnnotated: ['Test', 'ParameterizedTest', 'RepeatedTest', 'TestFactory']
```

We use a those other 3 junit annotations that we need to add.

About the other rules we should talk about them in different issues. But, for example, I think that it's safe to disable `UndocumentedPublicClass` for every function in the test source set (not only the test functions).

**Comment by TWiStErRob** [comment]
> I think that it's safe to disable UndocumentedPublicClass for every function in the test source set (not only the test functions).

Uhm, no! Imagine classes like the ones listed [here](https://github.com/Kotlin/kotlinx.coroutines/issues/242), with 0 docs. But I see why we need separate issues :)

**Comment by marschwar** [comment]
I am also not a big fan of this. I think it is a good idea to change our detekt.yml but not the default config. I think detekt should remain framework agnostic and this would open the box of the Pandora .

**Comment by cortinico** [comment]
> I think detekt should remain framework agnostic and this would open the box of the Pandora .

+1 to this. I would stick to our principles 👍

**Comment by TWiStErRob** [comment]
If we're so strictly framework agnostic then maybe we should remove some of config mentioned in https://github.com/detekt/detekt/issues/5903#issuecomment-1516329910 ?

I'm always for sane defaults for common sense ([JUnit 4 510000](https://github.com/search?q=%22import+org.junit.Test%22+language%3AKotlin&type=code&l=Kotlin), [JUnit 5 133000](https://github.com/search?q=%22import+org.junit.jupiter.api.Test%22+language%3AKotlin&type=code&l=Kotlin)), with configurability for special cases (using Spek would be a special case, [v1 4600](https://github.com/search?q=%22import+org.jetbrains.spek.api.Spek%22+language%3AKotlin&type=code&l=Kotlin), [v2 8400](https://github.com/search?q=%22import+org.spekframework.spek2.Spek%22+language%3AKotlin&type=code&l=Kotlin)).

**Comment by BraisGabin** [comment]
Closing as there are 3 maintainers against. Thanks for the feedback!
